File: src/parsers/markdown/parser.ts
/**
 * Incremental Markdown Parser
 *
 * Parses Markdown tokens into Synth AST with support for incremental re-parsing.
 * Integrates with node pool and query index for maximum performance.
 */

import type { Tree, NodeId, BaseNode } from '../../types/index.js'
import { createTree, addNode } from '../../types/tree.js'
import type { Edit } from '../../core/incremental.js'
import { createIndex, type ASTIndex } from '../../core/query-index.js'
import { globalNodePool } from '../../core/node-pool.js'
import { IncrementalTokenizer } from './tokenizer.js'
import type { BlockToken } from './tokens.js'

/**
 * Incremental Markdown Parser
 *
 * Provides complete Markdown parsing with incremental re-parsing support.
 */
export class IncrementalMarkdownParser {
  private tokenizer: IncrementalTokenizer
  private tree: Tree | null = null
  private tokens: BlockToken[] = []
  private index: ASTIndex | null = null
  private source: string = ''

  constructor() {
    this.tokenizer = new IncrementalTokenizer()
  }

  /**
   * Full parse of Markdown text
   */
  parse(text: string): Tree {
    this.source = text

    // 1. Tokenize
    this.tokens = this.tokenizer.tokenize(text)

    // 2. Build AST
    this.tree = this.buildTree(this.tokens, text)

    // 3. Build index
    this.index = createIndex(this.tree)
    this.index.build()

    return this.tree
  }

  /**
   * Incremental parse after edit
   *
   * Strategy:
   * 1. Incremental tokenize
   * 2. Find affected nodes
   * 3. Re-parse affected region
   * 4. Structural sharing (reuse unchanged nodes)
   * 5. Rebuild index
   */
  parseIncremental(text: string, edit: Edit): Tree {
    if (!this.tree || !this.index) {
      throw new Error('Must call parse() before parseIncremental()')
    }

    this.source = text

    // 1. Incremental tokenize
    this.tokens = this.tokenizer.retokenize(text, edit, this.tokens)

    // 2. Find affected nodes using index
    const affectedNodeIds = this.findAffectedNodes(edit)

    // 3. Release affected nodes to pool
    for (const nodeId of affectedNodeIds) {
      const node = this.tree.nodes[nodeId]
      if (node) {
        globalNodePool.release(node)
      }
    }

    // 4. Re-build entire tree (for now - will optimize later)
    this.tree = this.buildTree(this.tokens, text)

    // 5. Rebuild index
    this.index = createIndex(this.tree)
    this.index.build()

    return this.tree
  }

  /**
   * Build AST from tokens
   */
  private buildTree(tokens: BlockToken[], source: string): Tree {
    const tree = createTree('markdown', source)

    // Convert each token to a node
    for (const token of tokens) {
      const nodeId = this.buildNode(tree, token, tree.root)
      if (nodeId !== null) {
        tree.nodes[tree.root]!.children.push(nodeId)
      }
    }

    return tree
  }

  /**
   * Build a single node from token
   */
  private buildNode(tree: Tree, token: BlockToken, parent: NodeId): NodeId | null {
    // Skip blank lines (don't create nodes for them)
    if (token.type === 'blankLine') {
      return null
    }

    switch (token.type) {
      case 'heading': {
        const headingId = addNode(tree, {
          type: 'heading',
          parent,
          children: [],
          span: {
            start: token.position.start,
            end: token.position.end,
          },
          data: { depth: token.depth },
        } as BaseNode)

        // Add text child
        const textId = addNode(tree, {
          type: 'text',
          parent: headingId,
          children: [],
          span: {
            start: token.position.start,
            end: token.position.end,
          },
          data: { value: token.text },
        } as BaseNode)

        tree.nodes[headingId]!.children.push(textId)
        return headingId
      }

      case 'paragraph': {
        const paragraphId = addNode(tree, {
          type: 'paragraph',
          parent,
          children: [],
          span: {
            start: token.position.start,
            end: token.position.end,
          },
        } as BaseNode)

        // Add text child
        const textId = addNode(tree, {
          type: 'text',
          parent: paragraphId,
          children: [],
          span: {
            start: token.position.start,
            end: token.position.end,
          },
          data: { value: token.text },
        } as BaseNode)

        tree.nodes[paragraphId]!.children.push(textId)
        return paragraphId
      }

      case 'codeBlock': {
        return addNode(tree, {
          type: 'code',
          parent,
          children: [],
          span: {
            start: token.position.start,
            end: token.position.end,
          },
          data: {
            value: token.code,
            lang: token.lang,
            meta: token.meta,
          },
        } as BaseNode)
      }

      case 'listItem': {
        const listItemId = addNode(tree, {
          type: 'listItem',
          parent,
          children: [],
          span: {
            start: token.position.start,
            end: token.position.end,
          },
          data: {
            checked: token.checked,
          },
        } as BaseNode)

        // Add text child
        const textId = addNode(tree, {
          type: 'text',
          parent: listItemId,
          children: [],
          span: {
            start: token.position.start,
            end: token.position.end,
          },
          data: { value: token.text },
        } as BaseNode)

        tree.nodes[listItemId]!.children.push(textId)
        return listItemId
      }

      case 'blockquote': {
        const blockquoteId = addNode(tree, {
          type: 'blockquote',
          parent,
          children: [],
          span: {
            start: token.position.start,
            end: token.position.end,
          },
        } as BaseNode)

        // Add text child
        const textId = addNode(tree, {
          type: 'text',
          parent: blockquoteId,
          children: [],
          span: {
            start: token.position.start,
            end: token.position.end,
          },
          data: { value: token.text },
        } as BaseNode)

        tree.nodes[blockquoteId]!.children.push(textId)
        return blockquoteId
      }

      case 'horizontalRule': {
        return addNode(tree, {
          type: 'thematicBreak',
          parent,
          children: [],
          span: {
            start: token.position.start,
            end: token.position.end,
          },
        } as BaseNode)
      }

      default: {
        console.warn(`Unknown token type: ${(token as any).type}`)
        return null
      }
    }
  }

  /**
   * Find affected nodes based on edit
   */
  private findAffectedNodes(edit: Edit): NodeId[] {
    if (!this.index) return []

    const affected = new Set<NodeId>()

    // Find nodes that overlap with the edit range
    for (let i = 0; i < this.tree!.nodes.length; i++) {
      const node = this.tree!.nodes[i]
      if (!node || !node.span) continue

      const nodeStart = node.span.start.offset
      const nodeEnd = node.span.end.offset

      // Check if node overlaps with edit range
      if (
        (nodeStart <= edit.startByte && nodeEnd >= edit.startByte) ||
        (nodeStart <= edit.oldEndByte && nodeEnd >= edit.oldEndByte) ||
        (nodeStart >= edit.startByte && nodeEnd <= edit.oldEndByte)
      ) {
        affected.add(i)

        // Also mark parent as affected
        if (node.parent !== null) {
          affected.add(node.parent)
        }
      }
    }

    return Array.from(affected)
  }

  /**
   * Get current tree
   */
  getTree(): Tree {
    if (!this.tree) {
      throw new Error('No tree available. Call parse() first.')
    }
    return this.tree
  }

  /**
   * Get current tokens
   */
  getTokens(): BlockToken[] {
    return this.tokens
  }

  /**
   * Get index
   */
  getIndex(): ASTIndex | null {
    return this.index
  }
}

/**
 * Create an incremental Markdown parser
 */
export function createMarkdownParser(): IncrementalMarkdownParser {
  return new IncrementalMarkdownParser()
}

/**
 * Parse Markdown text (one-shot)
 */
export function parseMarkdown(text: string): Tree {
  const parser = createMarkdownParser()
  return parser.parse(text)
}

File: crates/wasm-md/src/parser.rs
//! Markdown parser
//!
//! Converts tokenized Markdown into Synth AST.

use synth_wasm_core::{Node, Span, SynthResult, Tree};
use std::collections::HashMap;

use crate::tokenizer::{BlockToken, Tokenizer};

/// Markdown parser
pub struct MarkdownParser;

impl MarkdownParser {
    pub fn new() -> Self {
        Self
    }

    /// Parse Markdown source into a Tree
    pub fn parse(&self, source: &str) -> SynthResult<Tree> {
        let mut tree = Tree::new("markdown", source);
        let mut tokenizer = Tokenizer::new(source);
        let tokens = tokenizer.tokenize();

        for token in tokens {
            if let Some(node_id) = self.build_node(&mut tree, &token)? {
                tree.add_child(tree.root_id(), node_id)?;
            }
        }

        Ok(tree)
    }

    fn build_node(&self, tree: &mut Tree, token: &BlockToken) -> SynthResult<Option<u32>> {
        match token {
            BlockToken::BlankLine { .. } => Ok(None),

            BlockToken::Heading { depth, text, span } => {
                let mut data = HashMap::new();
                data.insert("depth".to_string(), serde_json::json!(*depth));

                let node = Node::new(0, "heading")
                    .with_span(*span)
                    .with_data(data);
                let heading_id = tree.add_node(node);

                // Add text child
                self.add_text_node(tree, heading_id, text, span)?;

                Ok(Some(heading_id))
            }

            BlockToken::Paragraph { text, span } => {
                let node = Node::new(0, "paragraph").with_span(*span);
                let para_id = tree.add_node(node);

                // Add text child
                self.add_text_node(tree, para_id, text, span)?;

                Ok(Some(para_id))
            }

            BlockToken::CodeBlock { lang, meta, code, span } => {
                let mut data = HashMap::new();
                if let Some(l) = lang {
                    data.insert("lang".to_string(), serde_json::json!(l));
                }
                if let Some(m) = meta {
                    data.insert("meta".to_string(), serde_json::json!(m));
                }
                data.insert("value".to_string(), serde_json::json!(code));

                let node = Node::new(0, "code")
                    .with_span(*span)
                    .with_data(data);
                let code_id = tree.add_node(node);

                Ok(Some(code_id))
            }

            BlockToken::ThematicBreak { span } => {
                let node = Node::new(0, "thematicBreak").with_span(*span);
                let id = tree.add_node(node);
                Ok(Some(id))
            }

            BlockToken::BlockQuote { text, span } => {
                let node = Node::new(0, "blockquote").with_span(*span);
                let quote_id = tree.add_node(node);

                // Add text child
                self.add_text_node(tree, quote_id, text, span)?;

                Ok(Some(quote_id))
            }

            BlockToken::ListItem { text, ordered, checked, span } => {
                let mut data = HashMap::new();
                data.insert("ordered".to_string(), serde_json::json!(*ordered));
                if let Some(c) = checked {
                    data.insert("checked".to_string(), serde_json::json!(*c));
                }

                let node = Node::new(0, "listItem")
                    .with_span(*span)
                    .with_data(data);
                let item_id = tree.add_node(node);

                // Add text child
                self.add_text_node(tree, item_id, text, span)?;

                Ok(Some(item_id))
            }
        }
    }

    fn add_text_node(&self, tree: &mut Tree, parent_id: u32, text: &str, span: &Span) -> SynthResult<()> {
        if text.is_empty() {
            return Ok(());
        }

        let mut data = HashMap::new();
        data.insert("value".to_string(), serde_json::json!(text));

        let node = Node::new(0, "text")
            .with_parent(parent_id)
            .with_span(*span)
            .with_data(data);

        let text_id = tree.add_node(node);
        tree.add_child(parent_id, text_id)?;

        Ok(())
    }
}

impl Default for MarkdownParser {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_heading() {
        let parser = MarkdownParser::new();
        let tree = parser.parse("# Hello World\n").unwrap();

        assert_eq!(tree.node_count(), 3); // root, heading, text
    }

    #[test]
    fn test_parse_multiple_blocks() {
        let parser = MarkdownParser::new();
        let tree = parser.parse("# Title\n\nParagraph text.\n").unwrap();

        // root, heading, text, paragraph, text
        assert!(tree.node_count() >= 3);
    }

    #[test]
    fn test_parse_code_block() {
        let parser = MarkdownParser::new();
        let tree = parser.parse("```rust\nfn main() {}\n```\n").unwrap();

        assert!(tree.node_count() >= 2); // root, code
    }
}
